src/Uuid/ModelUuid.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace Kirby\Uuid;
+
+use Kirby\Cms\App;
+use Kirby\Cms\Collection;
+
+/**
+ * Base for Uuids for models where id string
+ * is stored in the content, such as pages and files
+ *
+ * @package   Kirby Uuid
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+abstract class ModelUuid extends Uuid
+{
+	/**
+	 * @var \Kirby\Cms\ModelWithContent|null
+	 */
+	public Identifiable|null $model;
+
+	public function __construct(
+		string|null $uuid = null,
+		Identifiable|null $model = null,
+		Collection|null $context = null
+	) {
+		parent::__construct($uuid, $model, $context);
+
+		// ensure that ID gets generated right away if
+		// not yet stored any in content file
+		$this->id();
+	}
+
+	/**
+	 * Look up Uuid in local and global index
+	 * and return the identifiable model object.
+	 */
+	protected function findByIndex(): Identifiable|null
+	{
+		foreach ($this->indexes() as $model) {
+			if (static::retrieveId($model) === $this->id()) {
+				return $model;
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Returns the UUID's id string. If not set yet,
+	 * creates a neq unique ID and writes it to content file
+	 */
+	public function id(): string
+	{
+		if ($id = $this->uri->host()) {
+			return $id;
+		}
+
+		// generate ID and write to content file
+		$id = static::generate();
+
+		// make sure Kirby has the required permissions
+		// for the update action
+		$kirby = App::instance();
+		$user  = $kirby->auth()->currentUserFromImpersonation();
+		$kirby->impersonate('kirby');
+		$this->model = $this->model->update(['uuid' => $id]);
```

Comment:
This line is causing an endless loop and I'm not entirely sure why yet. It is solved by just using the underlying save method instead but we need to double-check if that has any side effects.

```suggestion
		$this->model = $this->model->save(['uuid' => $id]);
```